### ISSUE TITLE
Initial vector generation is broken in 3.9.0, causing rare failures

### ISSUE THREAD

**Issue by szhorvat** [opencollab/arpack-ng#410]
In ARPACK, [the `Xgetv0()` function](https://github.com/opencollab/arpack-ng/blob/master/SRC/dgetv0.f) function generates a random initial vector for a numerical iteration process. With low probability, it may occur that this random vector won't be suitable for the numerical algorithm. In this case, ARPACK re-generates it, up to three times, until a suitable one is found. This happens here:

https://github.com/opencollab/arpack-ng/blob/master/SRC/dsaitr.f#L409

Commit ce2e69a849da1d10dad5d6d3ec4db6120b3ecf50 breaks this process by making `Xgetv0()` deterministic. The only change the commit makes is to re-seed the random number generator with the very same fixed seed at the beginning of `Xgetv0()`. Therefore, if with this fixed seed `Xgetv0()` happens to produce an unsuitable "random" vector, all subsequent retries will obviously fail in the same way.

The fact that the above-linked code performs random re-tries until success makes it obvious that ce2e69a849da1d10dad5d6d3ec4db6120b3ecf50 (which removes the randomness) is incorrect.

----

I explained this already in detail in #401. This issue is to make the problem plain to see without having to go through the long discussion there. The description is kept intentionally simple.

----

While the failure occurs with low probability, and it is dependent on the LAPACK implementation, it does occur in practice. It is currently causing the igraph test suite to fail with ARPACK 3.9.0 on several platforms ([see e.g. with msys2](https://github.com/igraph/igraph/actions/runs/4388756464/jobs/7685610006)). I confirmed that the failure mode is precisely the one described above, by adding `print` statements to follow the code path, and verify that inside `Xgetv0()` the very same "random" vector keeps getting generated again and again. A direct revert of ce2e69a849da1d10dad5d6d3ec4db6120b3ecf50 eliminates the problem.

**Comment by fghoussen** [comment]
Would be good, if possible, to have reproducible / deterministic data to reproduce the problem.

**Comment by szhorvat** [comment]
Can you address the obvious logic flaw that I described above?

**Comment by szhorvat** [comment]
This issue should be obvious from following the logic of the code. It would be good if you could clarify which part of the description needs clarification, and what sort of reproducer will provide this.

As I said above, part of ARPACK re-tries a certain operation with several random vectors, until it succeeds. In the vast majority of cases, the very first try will succeed. I am assuming that you want to see a reproducer that demonstrates that the first try may fail. I provide this below. It is based on the igraph test case failure described in #401. Note that this only work when linking against certain BLAS/LAPACK on certain platforms, [see here](https://github.com/opencollab/arpack-ng/issues/401#issuecomment-1440558326).

https://gist.github.com/szhorvat/f1d06fa8e58901d17cd3e06eac8891f5

After ce2e69a849da1d10dad5d6d3ec4db6120b3ecf50, DSAUPD() fails with `INFO = -9999` in this program. You will see output similar to the following:

```
Trial 0:
Finished.
INFO = 0

Trial 1:
Finished.
INFO = 0

Trial 2:
Finished.
INFO = 0

Trial 3:
Finished.
INFO = 0

Trial 4:
Finished.
INFO = 0

Trial 5:
Finished.
INFO = 0

Trial 6:
Finished.
INFO = -9999
```

**Comment by szhorvat** [comment]
For example, with this change:

```diff
Last login: Sun Mar 12 16:53:43 on ttys007
contWelcome to fish, the friendly interactive shell
~/R/arpack-example $ port contents arpack
Port arpack contains:
  /opt/local/include/arpack-ng/arpack.h
  /opt/local/include/arpack-ng/arpack.hpp
  /opt/local/include/arpack-ng/arpackdef.h
  /opt/local/include/arpack-ng/arpackicb.h
  /opt/local/include/arpack-ng/debug.h
  /opt/local/include/arpack-ng/debugF90.h
  /opt/local/include/arpack-ng/debug_c.h
  /opt/local/include/arpack-ng/debug_c.hpp
  /opt/local/include/arpack-ng/stat.h
  /opt/local/include/arpack-ng/statF90.h
  /opt/local/include/arpack-ng/stat_c.h
  /opt/local/include/arpack-ng/stat_c.hpp
  /opt/local/lib/libarpack.2.dylib
  /opt/local/lib/libarpack.dylib
  /opt/local/lib/pkgconfig/arpack.pc
  /opt/local/lib/pkgconfig/arpackSolver.pc
  /opt/local/lib/pkgconfig/parpack.pc
~/R/arpack-example $ cd Repos/macports-ports/^C
~/R/arpack-example $ cd ~/Repos/arpack-ng/build/
~/R/a/build (master|✚3) $ cmake ..
-- Fortran compiler does support iso_c_binding.
-- Configuration summary for arpack-ng-3.9.0:
   -- prefix: /Users/szhorvat/apl
   -- MPI: OFF (ICB provided )
   -- ICB: ON
   -- INTERFACE64: 0
   -- FC:      /opt/local/bin/gfortran
   -- FCFLAGS: -O3 -DNDEBUG -O3 
   -- CC:      /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
   -- CCFLAGS: -O3 -DNDEBUG 
   -- CXX:      /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
   -- CXXFLAGS: -O3 -DNDEBUG 
   -- BLAS:
      -- link:    /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Accelerate.framework
   -- LAPACK:
      -- link:    /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Accelerate.framework
      -- link:    -lm
      -- link:    -ldl
-- Configuring done (0.6s)
-- Generating done (0.0s)
-- Build files have been written to: /Users/szhorvat/Repos/arpack-ng/build
~/R/a/build (master|✚3) $ ninja install
[0/1] Install the project...
-- Install configuration: "Release"
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.2.1.0.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.2.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/pkgconfig/arpack.pc
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debugF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/statF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackdef.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackicb.h
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets-release.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config-version.cmake
~/R/a/build (master|✚3) $ ninja install
[15/16] Install the project...
-- Install configuration: "Release"
-- Installing: /Users/szhorvat/apl/lib/libarpack.2.1.0.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.2.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/pkgconfig/arpack.pc
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debugF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/statF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackdef.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackicb.h
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets-release.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config-version.cmake
~/R/a/build (master|✚3) $ ninja install
[17/18] Install the project...
-- Install configuration: "Release"
-- Installing: /Users/szhorvat/apl/lib/libarpack.2.1.0.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.2.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/pkgconfig/arpack.pc
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debugF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/statF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackdef.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackicb.h
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets-release.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config-version.cmake
~/R/a/build (master|✚3) $ ninja install
[3/16] Building Fortran object CMakeFiles/arpack.dir/SRC/dgetv0.f.o
FAILED: CMakeFiles/arpack.dir/SRC/dgetv0.f.o 
/opt/local/bin/gfortran -I/Users/szhorvat/Repos/arpack-ng/SRC -I/Users/szhorvat/Repos/arpack-ng/build -I/Users/szhorvat/Repos/arpack-ng -I/Users/szhorvat/Repos/arpack-ng/ICB -O3 -DNDEBUG -O3 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.14 -fPIC -fpreprocessed -c CMakeFiles/arpack.dir/SRC/dgetv0.f-pp.f -o CMakeFiles/arpack.dir/SRC/dgetv0.f.o
/Users/szhorvat/Repos/arpack-ng/SRC/dgetv0.f:228:23:

  228 |             print *, "generating random resid in dgetv0(), resid is now:"
      |                       1
Error: Unterminated character constant beginning at (1)
ninja: build stopped: subcommand failed.
~/R/a/build (master|✚3) [1] $ ninja install
[15/16] Install the project...
-- Install configuration: "Release"
-- Installing: /Users/szhorvat/apl/lib/libarpack.2.1.0.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.2.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/pkgconfig/arpack.pc
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debugF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/statF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackdef.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackicb.h
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets-release.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config-version.cmake
~/R/a/build (master|✚3) $ ninja install
[3/16] Building Fortran object CMakeFiles/arpack.dir/SRC/dgetv0.f.o
FAILED: CMakeFiles/arpack.dir/SRC/dgetv0.f.o 
/opt/local/bin/gfortran -I/Users/szhorvat/Repos/arpack-ng/SRC -I/Users/szhorvat/Repos/arpack-ng/build -I/Users/szhorvat/Repos/arpack-ng -I/Users/szhorvat/Repos/arpack-ng/ICB -O3 -DNDEBUG -O3 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.14 -fPIC -fpreprocessed -c CMakeFiles/arpack.dir/SRC/dgetv0.f-pp.f -o CMakeFiles/arpack.dir/SRC/dgetv0.f.o
/Users/szhorvat/Repos/arpack-ng/SRC/dgetv0.f:228:23:

  228 |             print *, 'generating random resid in dgetv0(), resid is now: '
      |                       1
Error: Unterminated character constant beginning at (1)
ninja: build stopped: subcommand failed.
~/R/a/build (master|✚3) [1] $ ninja install
[15/16] Install the project...
-- Install configuration: "Release"
-- Installing: /Users/szhorvat/apl/lib/libarpack.2.1.0.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.2.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/pkgconfig/arpack.pc
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debugF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/statF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackdef.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackicb.h
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets-release.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config-version.cmake
~/R/a/build (master|✚3) $ ninja install
[15/16] Install the project...
-- Install configuration: "Release"
-- Installing: /Users/szhorvat/apl/lib/libarpack.2.1.0.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.2.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/pkgconfig/arpack.pc
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debugF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/statF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackdef.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackicb.h
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets-release.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config-version.cmake
~/R/a/build (master|✚3) $ ninja install
[0/1] Install the project...
-- Install configuration: "Release"
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.2.1.0.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.2.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/libarpack.dylib
-- Up-to-date: /Users/szhorvat/apl/lib/pkgconfig/arpack.pc
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpack.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debugF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/debug_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/statF90.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/stat_c.hpp
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackdef.h
-- Up-to-date: /Users/szhorvat/apl/include/arpack/arpackicb.h
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackngTargets-release.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config.cmake
-- Up-to-date: /Users/szhorvat/apl/lib/cmake/arpackng/arpackng-config-version.cmake
~/R/a/build (master|✚3) $ cd ..
~/R/arpack-ng (master|✚3) $ git diff
diff --git a/SRC/dgetv0.f b/SRC/dgetv0.f
index 8be4fa2..f3d473d 100644
--- a/SRC/dgetv0.f
+++ b/SRC/dgetv0.f
@@ -225,6 +225,8 @@ c
          if (.not.initv) then
             idist = 2
             call dlarnv (idist, iseed, n, resid)
+            print *, "generating random resid in dgetv0(), resid is now"
+            print *, resid
          end if
 c
 c        %----------------------------------------------------------%
diff --git a/SRC/dsaitr.f b/SRC/dsaitr.f
index 3460d99..3902a14 100644
--- a/SRC/dsaitr.f
+++ b/SRC/dsaitr.f
@@ -406,6 +406,7 @@ c           | If in reverse communication mode and |
 c           | RSTART = .true. flow returns here.   |
 c           %--------------------------------------%
 c
+            print *, "calling dgetv0() from dsaitr(), itry = ", itry
             call dgetv0 (ido, bmat, itry, .false., n, j, v, ldv,
      &                   resid, rnorm, ipntr, workd, ierr)
             if (ido .ne. 99) go to 9000
diff --git a/SRC/dsaup2.f b/SRC/dsaup2.f
index fd4143f..42f5efb 100644
--- a/SRC/dsaup2.f
+++ b/SRC/dsaup2.f
@@ -324,6 +324,7 @@ c
    10 continue
 c
       if (getv0) then
+         print *, "calling dgetv0() from dsaup2"
          call dgetv0 (ido, bmat, 1, initv, n, 1, v, ldv, resid, rnorm,
      &                ipntr, workd, info)
 c
```

Output:

```
Trial 6:
 calling dgetv0() from dsaup2
 calling dgetv0() from dsaup2
 calling dgetv0() from dsaitr(), itry =            1
 generating random resid in dgetv0(), resid is now
  0.39574246391875789        8.6496039750016962E-004 -0.92272057899825910      -0.91656714952780050       0.11759638488413060      -0.29962625203712179       0.90382695702586346      -0.25045104802183715       0.33224741301423677      -0.29023922021963955     
 calling dgetv0() from dsaitr(), itry =            1
 calling dgetv0() from dsaitr(), itry =            2
 generating random resid in dgetv0(), resid is now
  0.39574246391875789        8.6496039750016962E-004 -0.92272057899825910      -0.91656714952780050       0.11759638488413060      -0.29962625203712179       0.90382695702586346      -0.25045104802183715       0.33224741301423677      -0.29023922021963955     
 calling dgetv0() from dsaitr(), itry =            1
 generating random resid in dgetv0(), resid is now
  0.39574246391875789        8.6496039750016962E-004 -0.92272057899825910      -0.91656714952780050       0.11759638488413060      -0.29962625203712179       0.90382695702586346      -0.25045104802183715       0.33224741301423677      -0.29023922021963955     
 calling dgetv0() from dsaitr(), itry =            1
 calling dgetv0() from dsaitr(), itry =            2
 generating random resid in dgetv0(), resid is now
  0.39574246391875789        8.6496039750016962E-004 -0.92272057899825910      -0.91656714952780050       0.11759638488413060      -0.29962625203712179       0.90382695702586346      -0.25045104802183715       0.33224741301423677      -0.29023922021963955     
 calling dgetv0() from dsaitr(), itry =            1
 generating random resid in dgetv0(), resid is now
  0.39574246391875789        8.6496039750016962E-004 -0.92272057899825910      -0.91656714952780050       0.11759638488413060      -0.29962625203712179       0.90382695702586346      -0.25045104802183715       0.33224741301423677      -0.29023922021963955     
 calling dgetv0() from dsaitr(), itry =            1
 calling dgetv0() from dsaitr(), itry =            2
 generating random resid in dgetv0(), resid is now
  0.39574246391875789        8.6496039750016962E-004 -0.92272057899825910      -0.91656714952780050       0.11759638488413060      -0.29962625203712179       0.90382695702586346      -0.25045104802183715       0.33224741301423677      -0.29023922021963955     
 calling dgetv0() from dsaitr(), itry =            3
 generating random resid in dgetv0(), resid is now
  0.39574246391875789        8.6496039750016962E-004 -0.92272057899825910      -0.91656714952780050       0.11759638488413060      -0.29962625203712179       0.90382695702586346      -0.25045104802183715       0.33224741301423677      -0.29023922021963955     
Finished.
INFO = -9999
```

Notice how after `itry > 1`, it generates a new "random" vector, but that vector is always the same due to ce2e69a849da1d10dad5d6d3ec4db6120b3ecf50. Thus if a failure occurs with the new vector, the failure will continue to occur in subsequent tries as well.

**Comment by fghoussen** [comment]
@szhorvat: can you PR some patch doing https://github.com/opencollab/arpack-ng/issues/401#issuecomment-1464899571 with for instance https://cyber.dabamos.de/programming/modernfortran/random-numbers.html (or any other way you like to change seed _according to bounds mentioned in dlarnv doc_: this is what ce2e69a fixes)? If seed change, v0 will change, right?

**Comment by szhorvat** [comment]
`dlarnv()` (or rather `dlaruv()`, which `dlarnv()` calls internally) appears to use `iseed` not merely as a seed, but as the _complete 48-bit state_ of its random number generation algorithm.

The usage pattern is like this: One puts in an `iseed` on the first call, then `dlarnv()` outputs a vector of random numbers _and updates `iseed`_. After this, the resulting `iseed` should be passed back at the next call of `dlarnv()`.

[There are two documented requirements for `iseed` in `dlarnv()`:](https://netlib.org/lapack/explore-html/df/dd1/group___o_t_h_e_rauxiliary_ga77e05a87ced667cbdb502aa87c72d056.html)
 - All elements must be in [0, 4095]. That's because `iseed` really represents 48 bits, stored as four 12-bit values.
 - The last element, `iseed(4)`, must be odd.

I don't see how these requirements could have been violated. On the _first_ call, they are clearly satisfied: `1`, `3`, `5`, `7` are fine (and the more complicated parallel case was also fine at the time of ce2e69a849da1d10dad5d6d3ec4db6120b3ecf50, as far as I can tell). In fact, ce2e69a849da1d10dad5d6d3ec4db6120b3ecf50 did not change the _initial_ values used in `iseed`.

On _subsequent_ calls, `iseed` is no longer the responsibility of ARPACK. One is supposed to use whatever the previous `dlarnv()` call returned. If that violates the requirements, that's a LAPACK bug, not an ARPACK issue. But such a violation sounds very unlikely. Nevertheless, I printed out the returned `iseed` values on my machine, and I did not see any problem.

Do you have an example of `dlarnv()` returning an invalid `iseed` with some LAPACK implementation? If so, which one, and can you show an example program that reproduces the invalid `iseed`?

**Comment by fghoussen** [comment]
> One puts in an iseed on the first call, then dlarnv() outputs a vector of random numbers and updates iseed.

So the intent of the original code and ce2e69a are both correct.

> Do you have an example of dlarnv() returning an invalid iseed with some LAPACK implementation?

Don't really remember. Long time ago: try MKL with (or without?) ILP64.
AFAIR iseed is uninitialized as the cumbersome `data` keyword did _not_ initialize init**s** https://github.com/opencollab/arpack-ng/blob/f36eb6c75dc686c8cebdaca0106422d77b04152d/SRC/dgetv0.f#L190

**Comment by szhorvat** [comment]
> > One puts in an iseed on the first call, then dlarnv() outputs a vector of random numbers and updates iseed.
> 
> So the intent of the original code and [ce2e69a](https://github.com/opencollab/arpack-ng/commit/ce2e69a849da1d10dad5d6d3ec4db6120b3ecf50) are both correct.

No, because after ce2e69a849da1d10dad5d6d3ec4db6120b3ecf50, you never pass back the state to `dlarnv()`. Instead, you always start from the same fixed state, thus `dlarnv()` returns the same "random" numbers every time.

> Don't really remember. Long time ago: try MKL with (or without?) ILP64.

Unfortunately I lost access to the system where I could test with the MKL.

If we don't have an example of the problem that ce2e69a849da1d10dad5d6d3ec4db6120b3ecf50 was supposed to fix, we can't move forward.



> AFAIR iseed is uninitialized as the cumbersome `data` keyword did _not_ initialize init**s**

Can you expand on this? I do not program in Fortran, so I searched for `data`, and all documentation I could find stated that `data` will set the initial value of variables. Then I tried the following program which gives the expected output.

```
      program test

      integer foo
      logical q, r

      data foo /137/
      data q /.true./
      data r /.false./

      write (*,*) 'foo = ', foo
      write (*,*) 'q = ', q, 'r = ', r

      stop
      end
```

Output:

```
 foo =          137
 q =  T r =  F
```

Once again, there needs to be a clear understanding of the problem that ce2e69a849da1d10dad5d6d3ec4db6120b3ecf50 was intending to fix. Otherwise I suggest directly reverting it, as it is clearly causing problems.

**Comment by fghoussen** [comment]
Did you try https://github.com/opencollab/arpack-ng/issues/401#issuecomment-1464899571?

>  Instead, you always start from the same fixed state

OK

> If we don't have an example of the problem

Look at issues attached to the ce2e69a PR. Try https://github.com/opencollab/arpack-ng/blob/396b90a0158f4ef4551b357974639d4ca38d765e/.github/workflows/jobs.yml#L156. 

> Can you expand on this?

No: not home + low connection / email access at the time.

>  clear understanding of the problem 

inits was not initialized (guess), so iseed wasn't neither (100% sure): get random iseed out of dlarnv bounds

**Comment by fghoussen** [comment]
@szhorvat: can you PR a revert? 

> If we don't have an example of the problem

Analyzing history, the problem appeared in this specific case: parpack + ILP64 + MKL. But, it's now out of reach #368.

**Comment by szhorvat** [comment]
I can do it the week after next

**Comment by fghoussen** [comment]
> I can do it the week after next

Sure! No hurry :)

**Comment by fghoussen** [comment]
Fixed by #423

**Comment by szhorvat** [comment]
@fghoussen Thanks! I'm looking forward to the release of 3.9.1.
